fs/fs.go

```diff
@@ -410,38 +421,107 @@ func (fs *filesystem) fetchSociIndex(ctx context.Context, imageRef, indexDigest,
 		return nil, err
 	}
 
-	var indexDesc ocispec.Descriptor
-	if indexDigest == "" {
-		log.G(ctx).Info("index digest not provided, making a Referrers API call to fetch list of indices")
-		imgDigest, err := digest.Parse(imageManifestDigest)
-		if err != nil {
-			return nil, fmt.Errorf("unable to parse image digest: %w", err)
-		}
+	indexDesc, err := fs.findSociIndexDesc(ctx, imageManifestDigest, indexDigest, remoteStore)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", snapshot.ErrNoIndex, err)
+	}
 
-		artifactClient := NewOCIArtifactClient(remoteStore)
+	log.G(ctx).WithField("digest", indexDesc.Digest.String()).Infof("fetching SOCI artifacts using index descriptor")
 
-		desc, err := artifactClient.SelectReferrer(ctx, ocispec.Descriptor{Digest: imgDigest}, defaultIndexSelectionPolicy)
-		if err != nil {
-			return nil, fmt.Errorf("%w: cannot fetch list of referrers: %w", snapshot.ErrNoIndex, err)
+	index, err := FetchSociArtifacts(ctx, refspec, indexDesc, fs.contentStore, remoteStore)
+	if err != nil {
+		return nil, fmt.Errorf("%w: error trying to fetch SOCI artifacts: %w", snapshot.ErrNoIndex, err)
+	}
+	return index, nil
+}
+
+func (fs *filesystem) findSociIndexDesc(ctx context.Context, imageManifestDigest string, sociIndexDigest string, remoteStore *orasremote.Repository) (ocispec.Descriptor, error) {
+	imgDigest, err := digest.Parse(imageManifestDigest)
+	if err != nil {
+		return ocispec.Descriptor{}, fmt.Errorf("unable to parse image digest: %w", err)
+	}
+
+	// 1. Try to find an explicit index digest
+	if sociIndexDigest != "" {
+		log.G(ctx).Debug("using provided soci index digest")
+		return parseIndexDigest(sociIndexDigest)
+	}
+	log.G(ctx).Debug("index digest not provided")
+
+	// 2. Try to find an index digest in the manifest labels if SOCI v2 is enabled.
+	if fs.pullmodes.SOCIv2.Enable {
+		log.G(ctx).Debug("checking for soci v2 index annotation")
+		desc, err := findSociIndexDescAnnotation(ctx, imgDigest, remoteStore)
```

Comment:
what if you encounter an error here? should we atleast log that?

Reply:
If the  error is not found, we do log it on line 458. Otherwise we return the error which will eventually get logged.